### PR TITLE
streamingest: add datadriven framework

### PR DESCRIPTION
**Pull request by adityamaru**
Every package deserves a datadriven driver. This change
introduces one for `streamingest` making it easier to write
tests without having to do a lot of setup. Note, the utility
methods were already very well modularized but this makes interacting
with them more intuitive and allows the author/reviewer to focus on
the actual test.

We expect this driver to keep evolving as more tests are written as
data driven tests. This is just meant to serve as a template.

Release note: None

Informs: CRDB-20208

### CODE REVIEW

**Comment by adityamaru** [comment]
I have a sneaky suspicion the stress failure is linked to https://github.com/cockroachdb/cockroach/pull/92825. Running with that patch seems to be fixing the failure. I am yet to confirm this but didn't want to hold up reviews on that.

**Review by lidorcarmel — APPROVED** [review]
looks great! only minor nits.

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -7349,6 +7378,7 @@ expires until the statement bundle is collected`,
 			Fn: func(ctx context.Context, evalCtx *eval.Context, args tree.Datums) (tree.Datum, error) {
 				ctx, sp := tracing.ChildSpan(ctx, "crdb_internal.fingerprint")
 				defer sp.Finish()
+				log.Infof(ctx, "starting fingerprint")
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/sql/sem/builtins/builtins.go:7381]: nope, debugging detritus.

[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/sql/sem/builtins/builtins.go:7381]: removed

File: pkg/ccl/streamingccl/streamingest/datadriven_test.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+					return nil
+				})
+
+			case "start-replication-stream":
+				ds.producerJobID, ds.replicationJobID = ds.replicationClusters.startStreamReplication()
+
+			case "wait-until-high-watermark":
+				var highWaterMark string
+				d.ScanArgs(t, "ts", &highWaterMark)
+				varValue, ok := ds.vars[highWaterMark]
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/datadriven_test.go:102]: The value of the `wait-until-high-watermark ts=$val` could be a var we have set before in the test via the `let` directive. And so we need to substitute the $var for its actual timestamp value here.

File: pkg/ccl/streamingccl/streamingest/datadriven_test.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+				d.ScanArgs(t, "as", &as)
+				output := ds.queryAs(t, as, d.Input)
+				output = strings.TrimSpace(output)
+				values := strings.Split(output, "\n")
+				if len(values) != len(d.CmdArgs)-1 {
+					t.Fatalf("expecting %d vars, found %d", len(d.CmdArgs)-1, len(values))
+				}
+				for i := range values {
+					key := d.CmdArgs[i].Key
+					if !strings.HasPrefix(key, "$") {
```
[the remark this answers is not in the document]
Comment (adityamaru) [reply pkg/ccl/streamingccl/streamingest/datadriven_test.go:129]: We have a check before this line that ensures that the number of $vars is the same as the number of rows outputted. So this should guard us against a mismatch. I reworked the logic a little bit to be more stable though.

**Review by lidorcarmel — APPROVED** [review]
lgtm

**Comment by adityamaru** [comment]
All failures are flakes being tracked elsewhere.

TFTR!
bors r=lidorcarmel

**Comment by adityamaru** [comment]
I think I need to move some testutils into testonly packages now that we disallow non-test cockroach files to import testcluster.

**Review by lidorcarmel — APPROVED** [review]

**Comment by adityamaru** [comment]
The flake is being actively looked at in https://github.com/cockroachdb/cockroach/issues/93731. Since this PR is conflict central for c2c code, I'm going to go ahead merging it.

bors r=lidorcarmel

**Comment by adityamaru** [comment]
bors r=lidorcarmel

**Pull request merged**